internal/security/secretstore/vmkencryption.go

```diff
@@ -0,0 +1,280 @@
+		wipeKey(keyShare) // Clear out binary version of encrypted key
+		wipeKey(nonce)    // Clear out nonce
+	}
+
+	initResp.EncryptedKeys = newKeys
+	initResp.Nonces = newNonces
+	initResp.Keys = nil       // strings are immutable, must wait for GC
+	initResp.KeysBase64 = nil // strings are immutable, must wait for GC
+	return nil
+}
+
+// DecryptInitResponse processes the InitResponse and decrypts the key shares
+// in the end, EncryptedKeys and Nonces are removed and replaced with
+// Keys and KeysBase64 in the resulting JSON like the init response was originally
+// Root token is left untouched
+func (v *VMKEncryption) DecryptInitResponse(initResp *secretstoreclient.InitResponse) error {
+
+	// This is a no-op unless LoadIKM() was called previously
+	if !v.encrypting {
+		return nil
+	}
+
+	newKeys := make([]string, len(initResp.EncryptedKeys))
+	newKeysBase64 := make([]string, len(initResp.EncryptedKeys))
+
+	for i, hexCiphertext := range initResp.EncryptedKeys {
+
+		hexNonce := initResp.Nonces[i]
+		nonce, err := hex.DecodeString(hexNonce)
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to decode hex bytes of nonce %s", err.Error()))
+			return err
+		}
+
+		cipherText, err := hex.DecodeString(hexCiphertext)
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to decode hex bytes of ciphertext"))
+			return err
+		}
+
+		keyShare, err := v.gcmDecryptKeyshare(cipherText, nonce, i) // Unwrap using a unique AES key
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to unwrap key %d: %s", i, err.Error()))
+			return err
+		}
+
+		newKeys[i] = hex.EncodeToString(keyShare)
+		newKeysBase64[i] = base64.StdEncoding.EncodeToString(keyShare)
+	}
+
+	initResp.Keys = newKeys
+	initResp.KeysBase64 = newKeysBase64
+	initResp.EncryptedKeys = nil
+	initResp.Nonces = nil
+	return nil
+}
+
+//
+// Internal methods
+//
+
+func (v *VMKEncryption) gcmEncryptKeyshare(keyshare []byte, counter int) ([]byte, []byte, error) {
+
+	info := fmt.Sprintf("vault%d", counter)
+
+	key, err := v.kdf.DeriveKey(v.ikm, aesKeyLength, info)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to derive encryption key for vault master key share %s", err.Error()))
+		return nil, nil, err
+	}
+	defer wipeKey(key)
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize block cipher %s", err.Error()))
+		return nil, nil, err
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize AES cipher %s", err.Error()))
+		return nil, nil, err
+	}
+
+	nonce := make([]byte, aesgcm.NonceSize())
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize random nonce %s", err.Error()))
+		return nil, nil, err
+	}
+
+	// Encrypt they key share, overwrite the input, and wipe the encrypted copy
+	// (input and output size are the same with AES)
+	ciphertext := aesgcm.Seal(nil, nonce, keyshare, nil)
+	wipeKey(keyshare)
```

Comment:
any reason not do do `defer wipeKey(keyshare)` at the top of this function too?

Reply:
removed to the top as a defer.